scipy/optimize/_lsq/common.py

```diff
@@ -0,0 +1,731 @@
+    """Compute J diag(d).
+
+    If `copy` is False, `J` is modified in place (unless being LinearOperator).
+    """
+    if copy and not isinstance(J, LinearOperator):
+        J = J.copy()
+
+    if issparse(J):
+        J.data *= d.take(J.indices, mode='clip')  # scikit-learn recipe.
+    elif isinstance(J, LinearOperator):
+        J = right_multiplied_operator(J, d)
+    else:
+        J *= d
+
+    return J
+
+
+def left_multiply(J, d, copy=True):
+    """Compute diag(d) J.
+
+    If `copy` is False, `J` is modified in place (unless being LinearOperator).
+    """
+    if copy and not isinstance(J, LinearOperator):
+        J = J.copy()
+
+    if issparse(J):
+        J.data *= np.repeat(d, np.diff(J.indptr))  # scikit-learn recipe.
+    elif isinstance(J, LinearOperator):
+        J = left_multiplied_operator(J, d)
+    else:
+        J *= d[:, np.newaxis]
+
+    return J
+
+
+def check_termination(dF, F, dx_norm, x_norm, ratio, ftol, xtol):
+    ftol_satisfied = dF < ftol * F and ratio > 0.25
+    xtol_satisfied = dx_norm < xtol * (xtol + x_norm)
+
+    if ftol_satisfied and xtol_satisfied:
+        return 4
+    elif ftol_satisfied:
+        return 2
+    elif xtol_satisfied:
+        return 3
+    else:
+        return None
+
+
+# Routines for robust loss function operations.
+
+
+IMPLEMENTED_LOSSES = ['linear', 'huber', 'soft_l1', 'cauchy', 'arctan']
+
+
+def compute_cost(f, loss, loss_scale):
+    if loss == 'linear':
+        return 0.5 * np.dot(f, f)
+
+    z = (f / loss_scale) ** 2
+    if loss == 'huber':
+        rho = z
+        rho[rho > 1] = 2 * z[rho > 1]**0.5 - 1
+    elif loss == 'soft_l1':
+        rho = 2 * ((1 + z)**0.5 - 1)
+    elif loss == 'cauchy':
+        rho = np.log1p(z)
+    elif loss == 'arctan':
+        rho = np.arctan(z)
+    elif callable(loss):
+        rho = loss(z)[0]
+    else:
+        raise ValueError("Unknown `loss`.")
+
+    return 0.5 * loss_scale**2 * np.sum(rho)
+
+
+def compute_loss_and_derivatives(f, loss, loss_scale):
+    if loss == 'linear':
+        return np.vstack((f**2, np.ones_like(f), np.zeros_like(f)))
+
+    rho = np.empty((3, f.size))
+    z = (f / loss_scale) ** 2
+
+    if loss == 'huber':
+        mask = z <= 1
+        rho[0, mask] = z[mask]
+        rho[0, ~mask] = 2 * z[~mask]**0.5 - 1
+        rho[1, mask] = 1
+        rho[1, ~mask] = z[~mask]**-0.5
+        rho[2, mask] = 0
+        rho[2, ~mask] = -0.5 * z[~mask]**-1.5
+    elif loss == 'soft_l1':
+        t = 1 + z
+        rho[0] = 2 * (t**0.5 - 1)
+        rho[1] = t**-0.5
+        rho[2] = -0.5 * t**-1.5
+    elif loss == 'cauchy':
+        rho[0] = np.log1p(z)
+        t = 1 + z
+        rho[1] = 1 / t
+        rho[2] = -1 / t**2
+    elif loss == 'arctan':
+        rho[0] = np.arctan(z)
+        t = 1 + z**2
+        rho[1] = 1 / t
+        rho[2] = -2 * z / t**2
+    elif callable(loss):
+        rho = np.atleast_2d(loss(z))
+    else:
+        raise ValueError("Unknown `loss`.")
```

Comment:
OK, this is a minor implementation detail, and this can be done separately, but still I guess I'd comment on this here anyway.  I personally really don't like these sorts of long if-elif-elif chains. This also looks wrong to branch over and over at each iteration, when the decision on which loss to use should actually be done once before the loop. And the duplication between `compute_loss_and_derivative` and `compute_cost` should be DRYed up.

I would do the following: at the module level, define normal functions

```
def arctan_loss(f, scale):
   z = f / scale
   out = np.empty((3, f.size))
   out[0] =  np.arctan(z)
   out[1] = ...
   out[2] = ...
   return out * scale**2
``` 
and so on. Maybe even allocate the `out` array only once and make these loss functions have signatures of `XYZ_loss(f, scale, out)`.

Also on the module level, define a dict mapping between string names and functions, 
`_loss_map = {'huber' : huber_loss, 'arctan': arctan_loss, ...}`

Then in `least_squares`, in the validation code, do 
```
try:
   loss_function = _loss_map[loss]
except KeyError:
   # linear loss or a user-supplied callable. Or handle the linear loss as a no-op function.
   # Validate that user callable here with known J0 and f0. Check that it actually has 
   # ndim=2, raise otherwise. 
```

And then in the main loop you have a validated `loss_function`, and you just use it.